[WEB-1420] 통합 스왑 구현

G-Gamja opened the pull request:
Squid swap api, 1inch swap api, Osmo swap을 하나의 페이지로 통합하여 통합 스왑 페이지를 구현했습니다.

### 체인 선택에 따라 사용할 스왑 api가 결정됩니다.
- 송, 수신 체인이 선택되면 사용할 api가 결정됩니다.
  - 송, 수신 체인이 Osmosis이면 Osmosis Swap을 사용합니다.
  - 송, 수신 체인이 EVM계열이며 같은 체인이라면 1inch Swap을 사용합니다.
  - 송, 수신 체인이 EVM계열이며 서로 다른 체인이거나, 송신 체인이 EVM이며 수신 체인이 Cosmos계열이라면 Squid Swap을 사용합니다.
  
- 체인이 선택되기 전까지는 토큰을 선택할 수 없습니다.

#### 스왑 페이지 라우팅
- 어떤 체인에서 스왑 페이지로 이동하느냐에 따라 송신체인이 자동 선택됩니다.
  - e.g 이더리움 체인에서 스왑 페이지로 이동 시 송신 체인이 이더리움으로 자동 선택됩니다.
  - 송신 체인 목록에 없는 체인에서 스왑 페이지로 이동 시에는 아무 체인도 선택되지 않은 상태로 이동합니다.

#### 체인 리스팅
- 송, 수신 어떤 체인을 선택하느냐에 따라 선택 가능한 Opposite 체인의 리스팅이 영향을 받습니다. 따라서 송, 수신 체인의 선 선택을 플래그로 설정하여 체인 선택 영향의 방향을 고정시켰습니다. 

#### 토큰 리스팅
- 몇몇 케이스에서 데이터 로딩에 지나치게 오랜 시간이 걸려 splice후 Intersection observer를 사용하여 스크롤할 수록 보여지는 토큰의 수를 늘렸습니다.

#### 1inch 스왑 시 특이점
- 1inch `Swap` api에서는 토큰의 price정보를 주지 않습니다. 이를 위해서는 spotPriceAggregator를 사용해서 가격 정보를 얻어야합니다.
  - 따라서 가격 정보를 노출시키지 않고 있으며, Native 코인만 Tx fee cost 노출을 위해 익스텐션 내에서 사용하는 코인 게코를 통해 가격 정보를 보여주고 있습니다.
  - 가격 정보가 없기에 price impact정보를 제거했습니다 
- 1inch 스왑에서 사용하는 스마트 컨트랙은 스왑할 토큰에 접근할 권한을 취득해야합니다. 따라서 스왑을 원하는 토큰이 1inch 컨트렉에 allowance가 설정되어 있지 않다면 이를 허락해주는 tx를 1번 post해야합니다.
  - 이때 설정되는 allowance의 정도는 최대 한도로 설정되어 있습니다.
  - allowance tx 사인 시에는 사인 후에 Swap 페이지로 라우팅 됩니다. 
- 1inch 스왑 사인 후에는 To Token을 커스텀 토큰으로 자동으로 추가하고 있습니다. 코인 게코 아이디는 1inch api에서 제공하지 않아 이는 포함하지 않고 있습니다.
- 현재는 `referrerAddress`를 할당하지 않고 있습니다. 추후 추천인 주소, fee 비율을 설정할 수 있습니다.
- 1inch Docs에서 api를 사용하여 나온 gaslimit를 그대로 사용하는 것을 추천하지 않고 있습니다. 따라서 디폴트 가스 배율을 적용시키고 있습니다.

#### Squid swap 시 특이점
- Swap tx를 post후 fee가 spike되는 경우 swap과정이 멈출 수 있습니다. 이러한 경우 axelar scan에서 직접 fee를 추가로 지불하거나, squid sdk에서 제공하는 fee추가 지불 메서드를 사용해야 합니다. 추후 구현될 activity페이지에서 구현될 가능성이 있습니다.

- 1inch 스왑 페이지
<img width="453" alt="스크린샷 2023-04-05 오후 4 38 47" src="https://user-images.githubusercontent.com/87967564/230013683-fcd85579-33f0-4e23-bce6-08ea741a44d5.png">
- squid 스왑 페이지
<img width="457" alt="스크린샷 2023-04-05 오후 4 40 07" src="https://user-images.githubusercontent.com/87967564/230015310-e45a52c6-3db5-402a-ac7f-ae55d625b444.png">
- osmo 스왑 페이지
<img width="444" alt="스크린샷 2023-04-05 오후 4 51 37" src="https://user-images.githubusercontent.com/87967564/230016569-5f987202-58a2-4da1-b601-f6e0affbe77c.png">

그 외
- Osmo, 1inch Swap시 볼 수 있는 Tx cost는 버튼으로 만들어 currecy, fee token amount 두 값을 변경하며 확인해 보실 수 있습니다.
- api마다 제공할 수 있는 정보가 달라 Swap Info를 각각에 알맞는 정보들을 선택적으로 노출시켰습니다.
<img width="405" alt="스크린샷 2023-04-05 오후 4 38 10" src="https://user-images.githubusercontent.com/87967564/230013545-fe1ec802-6b1d-4c4d-8998-6e2266992e73.png">
<img width="394" alt="스크린샷 2023-04-05 오후 4 38 14" src="https://user-images.githubusercontent.com/87967564/230013552-b95a8a5b-0e4c-45f0-8714-367ea651c6ee.png">

G-Gamja committed:
add squid sdk hook & 1inch api

G-Gamja committed:
add testing tx hash status & change init squid sdk style

G-Gamja committed:
building integrated skeleton & swap page refactoring & edit SubSideHeader

G-Gamja committed:
delete chalk

G-Gamja committed:
building chain listing & address selection

G-Gamja committed:
editing chain,token filtering

G-Gamja committed:
Merge branch 'develop' of https://github.com/cosmostation/cosmostation-chrome-extension into feature/WEB-1420-integrated-swap

G-Gamja committed:
edit chain filtertring data

G-Gamja committed:
add tokenBalaceSWR, token filtering & bottomList componentizing

G-Gamja committed:
AssetBottomList split & sorting token logic(added first)

G-Gamja committed:
adding token filtering, get token amount

G-Gamja committed:
delete comment

G-Gamja committed:
add osmo token, availableAmount logic

G-Gamja committed:
edit token vars name

G-Gamja committed:
integrating squid,osmo swap logic

G-Gamja committed:
adding 1inch routing & routing progress display

G-Gamja committed:
1inch allowance check & 1inch swap tx

G-Gamja committed:
adding squid exception, swap info(fee),swap Tx

G-Gamja committed:
adding Exception handling & revival swap button & integrate fee logic

G-Gamja committed:
revision swap coin,chain ux & add condition to avoid meaningless fetching squid route

G-Gamja committed:
delete meaningless balance decimals

G-Gamja committed:
Merge branch 'develop' of https://github.com/cosmostation/cosmostation-chrome-extension into feature/WEB-1420-integrated-swap

G-Gamja committed:
gas scaling & max button logic & improve osmo swap ux % restrict sllippage

G-Gamja committed:
add abi & squid token filtering & add token automatically when swap

G-Gamja committed:
delete useless comment

G-Gamja committed:
edit var name

G-Gamja committed:
add comment

G-Gamja committed:
Merge branch 'develop' of https://github.com/cosmostation/cosmostation-chrome-extension into feature/WEB-1420-integrated-swap

G-Gamja committed:
fix useless condition logic

G-Gamja committed:
fixing fee error logic & add allowance disable condtiton

G-Gamja committed:
apply review(auto token selecting...)&squid route fetching error(bridge fee)

G-Gamja committed:
edit i18 & swap info message & building infinite scroll

G-Gamja committed:
add Infinite scroll

G-Gamja committed:
Merge branch 'develop' of https://github.com/cosmostation/cosmostation-chrome-extension into feature/WEB-1420-integrated-swap

G-Gamja committed:
edit allowance fetching config & backDeQueue

G-Gamja committed:
edit squid error condition

G-Gamja committed:
delete unusable abi & edit coinGeckId field name & add error component

G-Gamja committed:
delete denom field in token type

G-Gamja committed:
Fix squid bridge fee display & replace token denom to address field

G-Gamja committed:
edit var names & allowance condition logic

G-Gamja committed:
fix squid bridge fee price display & edit tooltip arrow position & clearing

G-Gamja committed:
develop token searching ux & refactoring

G-Gamja committed:
componentize Intersect & osmoSwapMat,fee price currency button refactor var name

G-Gamja committed:
Merge branch 'develop' of https://github.com/cosmostation/cosmostation-chrome-extension into feature/WEB-1420-integrated-swap

G-Gamja committed:
add squidOnlyEVM chain & dimmed & edit chain comparing to id

G-Gamja committed:
apply support swap chain list& edit chain filtering & &name refactoring & currentSwapApi

G-Gamja committed:
apply OneInchRoute AxiosError type

G-Gamja committed:
add api icon button

G-Gamja committed:
edit api logo svg & squid receive cosmos token balance

G-Gamja committed:
edit api logo svg & add suspent to SupportSwapChainsSWR

G-Gamja committed:
edit api logo file name & add useSupportChainsSWR suspense

G-Gamja committed:
apply korean text on Swap button

ong-ar commented on src/Popup/pages/Wallet/OsmosisSwap/components/SlippageSettingDialog/index.tsx:
```diff
@@ -0,0 +1,117 @@
+  const { t } = useTranslation();
+
+  const [slippage, setCurrentSlippage] = useState(currentSlippage);
+  const [customSlippage, setCustomSlippage] = useState('');
```
state 하나로 가능할 것 같습니다.

G-Gamja replied:
커스텀입력과 프리셋 버튼과의 입력의 상태 차이를 두고자 기존과 같이 구현했습니다.

ong-ar commented on src/Popup/hooks/SWR/1inch/useAllowanceTxSWR.ts:
```diff
@@ -0,0 +1,28 @@
+};
+
+export function useAllowanceTxSWR(allowanceTxParam?: UseAllowanceTxSWRProps, config?: SWRConfiguration) {
+  const requestURL = allowanceTxParam && `${ONEINCH_BASE_URL}${allowanceTxParam.chainId}/approve/transaction?tokenAddress=${allowanceTxParam.tokenAddress}`;
```
`ONEINCH_BASE_URL` 마지막 `/` 지우고 `${ONEINCH_BASE_URL}/${allowanceTxParam.chainId}/approve/transaction?` 으로 변경해주세요.

G-Gamja replied:
수정했습니다

ong-ar commented on src/Popup/hooks/SWR/cosmos/useBalanceSWR.ts:
```diff
@@ -8,14 +8,14 @@ import { cosmosURL } from '~/Popup/utils/cosmos';
 
-  const address = accounts.data?.find((account) => account.id === chromeStorage.selectedAccountId)?.address[chain.id] || '';
-  const { getBalance } = cosmosURL(chain);
+  const address = (chain && accounts.data?.find((account) => account.id === chromeStorage.selectedAccountId)?.address[chain?.id]) || '';
```
요거 한번 정리 해야 할 것 같네요.
chain 이 없을 때 따로 처리하는 게 좋을 것 같습니다.

ong-ar commented on src/Popup/hooks/SWR/1inch/useTokenBalanceSWR.ts:
```diff
@@ -0,0 +1,56 @@
+  address: string;
+};
+
+export function useTokenBalanceSWR(currentEthereumNetwork?: IntegratedSwapEVMChain, token?: IntegratedSwapToken, config?: SWRConfiguration) {
```
이것도 기존거 같이 사용할 수 있을 것 같네요.

G-Gamja replied:
해당 메서드는 삭제한 후 기존 이더리움 토큰 밸런스를 가져오는 메서드에 입력 파라미터를 추가하는 방식으로 수정했습니다.

ong-ar commented on src/Popup/hooks/SWR/squid/useSquidSDKSWR.ts:
```diff
@@ -0,0 +1,29 @@
+
+import { SQUID_BASE_URL } from '~/constants/squid';
+
+export function useSquidSDKSWR(config?: SWRConfiguration) {
```
`useSquidSDKSWR` -> `useSquidSWR` 가 나을 것 같습니다.

G-Gamja replied:
수정했습니다

ong-ar commented on src/Popup/hooks/SWR/squid/useSquidRouteSWR.ts:
```diff
@@ -0,0 +1,36 @@
+      baseUrl: SQUID_BASE_URL,
+    });
+
+    await squid.init();
```
이것도 같이 쓸 수 있을 것 같네요.

G-Gamja replied:
수정했습니다

ong-ar commented on src/Popup/hooks/SWR/squid/useSquidTxStatusSWR.ts:
```diff
@@ -0,0 +1,27 @@
+      baseUrl: SQUID_BASE_URL,
+    });
+
+    await squid.init();
```
`useSquidSDKSWR` 를 이용해서 같이 사용할 수 있을 것 같네요

G-Gamja replied:
수정했습니다

ong-ar commented on src/Popup/hooks/osmoSwap/SWR/usePoolsSWR.ts:
```diff
@@ -6,7 +6,7 @@ import { get } from '~/Popup/utils/axios';
 
 export function usePoolSWR(poolId?: string, config?: SWRConfiguration) {
-  const requestURL = `https://lcd-osmosis-app.cosmostation.io/osmosis/gamm/v1beta1/pools/${poolId || ''}`;
+  const requestURL = poolId && `https://lcd-osmosis-app.cosmostation.io/osmosis/gamm/v1beta1/pools/${poolId}`;
```
이건 왜 바꾼겁니까?

G-Gamja replied:
코드 가독성이 좋아보일 것이라 생각했는데 리턴 타입이 2개가 되는 상황을 예상하지 못했습니다. 기존 코드로 다시 수정했습니다

ong-ar reviewed (changes requested):
볼게 너무 많아서 내일 한번 같이 보시죠

G-Gamja commented:
> 볼게 너무 많아서 내일 한번 같이 보시죠

넵 알겠습니다

ong-ar commented on src/Popup/hooks/useCurrent/useCurrentQueue.ts:
```diff
@@ -24,12 +24,26 @@ export function useCurrentQueue() {
     return queues.length > 0 ? queues[0] : null;
   };
 
+  const backDeQueue = async () => {
```
중복 코드 제거

G-Gamja replied:
해당 메서드 삭제한 후에 기존 메서드에 path 파라미터를 옵셔널로 받을 수 있도록 수정했습니다

ong-ar commented on src/Popup/pages/Account/Management/entry.tsx:
```diff
@@ -49,7 +49,7 @@ export default function Entry() {
   return (
     <Container>
-      <SubSideHeader title={t('pages.Account.Management.layout.title')} onClick={() => navigateBack()}>
+      <SubSideHeader title={t('pages.Account.Management.layout.title')}>
```
다시 살려주세요.

G-Gamja replied:
수정했습니다

ong-ar commented on src/Popup/pages/Popup/Ethereum/Transaction/entry.tsx:
```diff
@@ -230,6 +237,43 @@ export default function Entry({ queue }: EntryProps) {
     currentEthereumNetwork.chainId,
   ]);
 
+  const provider = new Web3.providers.HttpProvider(currentEthereumNetwork.rpcURL, {
```
전체적으로 다시 determineTxType 을 이용해서 abi 풀어서 해보세요.

G-Gamja replied:
kind타입을 추가하여 determinTxType으로 erc20과 swap(1inch)를 구분할 수 있도록 수정했습니다.

G-Gamja committed:
edit hooks url optioning & change tokenbalance hook to origin & add network param on TokenBalance hook &change OsmosisSwap page translation route

G-Gamja committed:
edit IntegratedSwapToken type

G-Gamja committed:
restore SubSideHeader & edit hook url optioning & add param in dequeue() & OneInchSwapTxMessage & determinTxType()

G-Gamja committed:
edit squidSDK constant name

G-Gamja committed:
Merge branch 'develop' of https://github.com/cosmostation/cosmostation-chrome-extension into feature/WEB-1420-integrated-swap

G-Gamja committed:
edit determineTxType() & delete useless component

G-Gamja committed:
apply memoization & catch minor errors

G-Gamja committed:
edit translation

ong-ar commented on src/Popup/pages/Wallet/components/cosmos/NativeChainCard/index.tsx:
```diff
@@ -534,7 +534,7 @@ export default function NativeChainCard({ chain, isCustom = false }: NativeChain
         </Button>
         {chain.id === OSMOSIS.id && (
-          <IconButtonContainer>
+          <IconButtonContainer sx={{ display: 'none' }}>
```
다시 살려주세요.

G-Gamja replied:
수정했습니다

G-Gamja committed:
add EthereumContractKind & limit swap input & restore osmo swap page

G-Gamja committed:
add default swap assets selecting logi & restore max button

ong-ar commented on src/Popup/pages/Popup/Ethereum/Transaction/entry.tsx:
```diff
@@ -593,8 +607,20 @@ export default function Entry({ queue }: EntryProps) {
                         }
-
-                        await deQueue();
+                        if (queue.channel === 'inApp' && txType.data?.type === 'approve') {
```
erc20 특정 시켜주세요.

G-Gamja replied:
수정은 완료했습니다. 현재 1inch api사용이 블락되어서 테스트를 못하고 있습니다. 테스트 후에 리졸브하겠습니다.

G-Gamja committed:
delete miss assiging contract kind in determineTxType

ong-ar commented on src/constants/route.ts:
```diff
@@ -6,6 +6,7 @@ export const PATH = {
   WALLET: '/wallet',
   WALLET__SEND: '/wallet/send',
   WALLET__SWAP: '/wallet/swap',
+  WALLET__OSMOSIS_SWAP: '/wallet/osmosisSwap',
```
`osmosisSwap` -> `osmosis-swap` 변경해주세요.

G-Gamja replied:
수정했습니다

G-Gamja committed:
display unAvailable Chain item & edit osmosis swap route key

G-Gamja committed:
check allowance tx contract kind

G-Gamja committed:
add useTokensBalanceSWR & sort tokens by balance

G-Gamja committed:
Merge branch 'develop' of https://github.com/cosmostation/cosmostation-chrome-extension into feature/WEB-1420-integrated-swap

G-Gamja committed:
edit 1inch api url

G-Gamja committed:
add useSquidTokensSWR

ong-ar commented on src/Popup/pages/Wallet/Swap/entry.tsx:
```diff
@@ -1,361 +1,920 @@
+    }
 
-  const swapFeeRate = useMemo(() => poolData.data?.pool.pool_params.swap_fee || '0', [poolData.data?.pool.pool_params.swap_fee]);
+    if (currentSwapAPI === 'squid' && currentToChain?.line === ethereumChain.line) {
```
`ethereumChain.line` -> `ETHEREUM`

G-Gamja committed:
feat: edit squid fee calculation logic

- edit squid source,cross chain fee calculation
- edit ChainFee component
- exclude "Express fee" in cross chain fee

G-Gamja committed:
delete ethereumChain const

ong-ar reviewed (approved):
고생하셨습니다.
